src/event_sources/epoll_linux.cc

```diff
@@ -86,39 +80,35 @@ EpollEventSource::EpollEventSource() : EventSource("Epoll")
     FATAL("failed to register close fd: %d\n", errno);
   }
 
-  spawn();
+  return spawn();
 }
 
-EpollEventSource::~EpollEventSource() {
+void EpollEventSourceBase::stop() {
   close(control_write_);
```

Comment:
Maybe reset `control_write_` and `epoll_fd_`? You could even check that they have been reset in `start`.